README.md
# AP2000-Sensor
 Small class to request/receive/process data from the Aquaread AP-2000 sensor


## Details

- This code was originally intended to run on an arduino MKR WAN 1300 board
- The code was made using the PlatformIO extension on Visual Studio Code
- This code have been made using Arduino framework

## Dependencies

- envirodiy/SDI-12 version 2.1.4
- arduino-libraries/MKRWAN version 1.1.0

## Notes

- Don't forget to edit the platformio.ini in case of any hardware or dependencies modification
- Make sure the used LoRa board/module you use is compatible with the MKR WAN Arduino's library
- This repository will probably never be updated, use with caution
- Fill appEui and appKey (in main.cpp) with your own you get on TTN after creating an app
- To see the debug prints on the serial define _DEBUG

CHANGELOG.md
# Change Log

## [0.0.2] - 2022-03-31

### Added

- Sleep mode between two aquisitions
- Reset function

### Changed

- Reboot after a bit on critical error

## [0.0.1] - 2022-03-25

### Changed

- Puts everything back to English
- Debug prints will only be displayed if _DEBUG is defined

## [0.0.0] - 2022-03-25

### Changed

- Cleanup

### Added

- Comments


src/main.cpp
/*!
 * @file main.cpp
 *
 * @section intro_sec Introduction
 *
 * This is a program to read and send to TTN AP-2000 sensor's measurements via LoRaWAN.
 *
 * @section author Author
 *
 * Written by The Last Resort.
 */


#include <Arduino.h>

#include "LowPower.h"
#include "RTCZero.h"
#include <SDI12.h>
#include <MKRWAN.h>
#include "ap2000.hpp"

#define RESET() digitalWrite(RESET_PIN, LOW)

constexpr uint32_t PORT_SERIE = 115200;
constexpr uint8_t DATA_PIN = 12;
constexpr uint8_t WAKE_UP_PIN = 6;
constexpr uint8_t RESET_PIN = 12;


LoRaModem modem;
Ap2000Sensor sensor(DATA_PIN, 0);
const String appEui = "";
const String appKey = "";

RTCZero rtc;

void alarm() 
{
  return;
}

void setup()
{
  Serial.begin(PORT_SERIE);

  // digitalWrite(10, HIGH);    // Power consumption test
  // pinMode(10, OUTPUT);

  digitalWrite(RESET_PIN, HIGH);
  pinMode(RESET_PIN, OUTPUT);

  rtc.begin();
  sensor.SensorBegin();

  if (!modem.begin(EU868)) // Setup LoRa
  { 
    #ifdef _DEBUG
      Serial.println("Failed to start module");
    #endif
    rtc.setDate(0,0,0);
    rtc.setTime(0,0,0);
    rtc.setAlarmTime(0,2,0);
    rtc.enableAlarm(rtc.MATCH_HHMMSS);
    rtc.attachInterrupt(alarm);
    LowPower.standby();
    RESET();
  }

  Serial.println(modem.version());
  Serial.println(modem.deviceEUI());
  int connected = modem.joinOTAA(appEui, appKey);

  if (!connected)
  {
    #ifdef _DEBUG
      Serial.println("A problem has occured while joining the TTN app");
    #endif
    rtc.setDate(0,0,0);
    rtc.setTime(0,0,0);
    rtc.setAlarmTime(0,2,0);
    rtc.enableAlarm(rtc.MATCH_HHMMSS);
    rtc.attachInterrupt(alarm);
    LowPower.standby();
    RESET();
  }
  
  pinMode(WAKE_UP_PIN, OUTPUT);
}

void loop()
{
  digitalWrite(WAKE_UP_PIN, HIGH); // Wake up the black box
  
  delay(40000);                   // Data won't be ready and stable until 40s

  #ifdef _DEBUG
    sensor.DispSensorInfos(Serial); // Debug test
  #endif

  SensorMeasurements *sm = new SensorMeasurements; // Get measurments
  sensor.RequestMeasurementsAndWaitUntilReady();
  sensor.GetSensorMeasurements(sm, 16, false);


  float tmp = sm->temperature;
  float oxy = sm->dissolvedOxygenSaturation;
  float sal = sm->salinity;
  float tur = sm->turbidity;
  float ph = sm->ph;
  delete sm;

  uint16_t tTmp = tmp * 100; // Data formatting for sending through LoRa
  uint16_t tOxy = oxy * 100;
  uint16_t tSal = sal * 100;
  uint16_t tTur = tur * 100;
  uint16_t tPh = ph * 100;

  #ifdef _DEBUG
    Serial.println("temperature: " + String(tmp) + "°C"); // Debug output
    Serial.println("oxygene: " + String(oxy) + "%");
    Serial.println("salinite: " + String(sal) + "g/L");
    Serial.println("turbitité: " + String(tur) + "FNU");
    Serial.println("ph: " + String(ph));
    Serial.println("\n___________________________________\n");
  #endif

  byte payload[10]; // Payload construction
  payload[0] = highByte(tTmp);
  payload[1] = lowByte(tTmp);
  payload[2] = highByte(tOxy);
  payload[3] = lowByte(tOxy);
  payload[4] = highByte(tSal);
  payload[5] = lowByte(tSal);
  payload[6] = highByte(tTur);
  payload[7] = lowByte(tTur);
  payload[8] = highByte(tPh);
  payload[9] = lowByte(tPh);

  modem.beginPacket(); // Sending data via LoRa
  modem.write(payload, sizeof(payload));
  int err = modem.endPacket(true);
  #ifdef _DEBUG
    if (err > 0)
    {
      Serial.println("Message has been sent successfully");
      Serial.println();
    }
    else
    {
      Serial.println("Error while sending the message");
    }
  #endif

  digitalWrite(WAKE_UP_PIN, LOW); // Put the blackbox back to sleep
  // Serial.println("Going to sleep");      // Power consumtion test
  // digitalWrite(10, HIGH);
  // delay(10000);
  // digitalWrite(10, LOW);

  // rtc.setDate(0,0,0);
  // rtc.setTime(0,0,0);
  // rtc.setAlarmTime(0,0,20);
  // rtc.enableAlarm(rtc.MATCH_HHMMSS);
  // rtc.attachInterrupt(alarm);
  // LowPower.standby();
  // delay(10);

  // Serial.println("woke up");
  // digitalWrite(10, HIGH);
  // delay(3000);
  // digitalWrite(10, LOW);
  // delay(3000);
}


src/ap200.cpp
#include "ap200.hpp"


/*!
 *  @brief  Instantiates a new Ap2000Sensor class
 *  @param  sdiPin
 *          Pin number that sensor is connected
 *  @param  sensorAddress
 *          Address of the sensor
 */
Ap2000Sensor::Ap2000Sensor(uint8_t sdiPin, uint8_t sensorAddress)
{
    sensorAddr = sensorAddress;
    sensor = new SDI12(sdiPin);
}

/*!
 *  @brief  Destructor of the Ap2000Sensor class
 */
Ap2000Sensor::~Ap2000Sensor()
{
    delete sensor;
}

/*!
 *  @brief  Initiate the SDI-12 communication
 */
void Ap2000Sensor::SensorBegin()
{
    sensor->begin();
}

/*!
 *  @brief  Get the sensor general infos
 *
 *	@return Return all the general sensor infos
 */
SensorInfos Ap2000Sensor::GetSensorInfos()
{
    SensorInfos si;
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "I!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    si.addr = sensorReply.substring(0, 1).toInt();
    si.sdiVersion = sensorReply.substring(1, 3).toFloat() / 10;
    sensorReply.substring(3, 11).toCharArray(si.vendorId, 8);
    si.vendorId[8] = '\0';
    sensorReply.substring(11, 17).toCharArray(si.model, 6);
    si.model[6] = '\0';
    si.fwVersion = sensorReply.substring(17, 20).toFloat() / 100;
    sensorReply.substring(20).toCharArray(si.serialNumber, 13);
    si.serialNumber[13] = '\0';
    sensor->clearBuffer();

    return si;
}

/*!
 *  @brief  Get the sensor general infos and display them
 *  @param  serial
 *          Serial for framework that uses HardwareSerial
 */
void Ap2000Sensor::DispSensorInfos(HardwareSerial &serial)
{
    SensorInfos si = GetSensorInfos();

    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Display the sensor general infos provided
 *  @param  serial
 *          Serial for framework that uses HardwareSerial
 *  @param  si
 *          SensorInfos structure to display
 */
void Ap2000Sensor::DispSensorInfos(HardwareSerial &serial, SensorInfos &si)
{
    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Get the sensor general infos and display them
 *  @param  serial
 *          Serial for framework that uses Serial_
 */
void Ap2000Sensor::DispSensorInfos(Serial_& serial) {
    SensorInfos si = GetSensorInfos();

    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Ping as SDI-12 address
 *  @param  addr
 *          Address of the sensor to ping
 *	@return Return true if the sensor replied to the ping otherwise return false
 */
bool Ap2000Sensor::IsSensorActive(uint8_t addr)
{
    String cmd = (String)addr + '!';
    for (int i = 0; i < 3; i++)
    {
        sensor->sendCommand(cmd);
        delay(100);
        if (sensor->available())
        {
            sensor->clearBuffer();
            return true;
        }
    }
    sensor->clearBuffer();
    return false;
}

/*!
 *  @brief  Request measurments to the sensor
 *  @param  mInfos
 *          char[2] that get filled with the time to wait until the measurements are ready and how many there is
 */
void Ap2000Sensor::RequestMeasurements(uint8_t *mInfos)
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    mInfos[0] = sensorReply.substring(1, 4).toInt();
    mInfos[1] = sensorReply.substring(4).toInt();
    sensor->clearBuffer();
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 *  @param  mInfos
 *          char[2] that get filled with the time to wait until the measurements are ready and how many there is
 */
void Ap2000Sensor::RequestMeasurementsAndWaitUntilReady(uint8_t *mInfos)
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    mInfos[0] = sensorReply.substring(1, 4).toInt();
    mInfos[1] = sensorReply.substring(4).toInt();
    sensor->clearBuffer();
    delay(mInfos[0] * 1000);
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 */
void Ap2000Sensor::RequestMeasurementsAndWaitUntilReady()
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();
    sensor->clearBuffer();

    delay(sensorReply.substring(1, 4).toInt() * 1000);
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 *  @param  measurements
 *          Structure that will be provisioned with the measurements received
 *  @param  resultsExpected
 *          The number of measurements expected
 *  @param  requestAgain
 *          Force a new measurement request before getting them
 */
void Ap2000Sensor::GetSensorMeasurements(SensorMeasurements* measurements, uint8_t resultsExpected, bool requestAgain)
{
    if (!measurementRequested || !requestAgain)
        this->RequestMeasurementsAndWaitUntilReady();

    uint8_t resultsReceived = 0;
    float* mIndex = (float*)measurements;
    uint8_t cmd_number = 0;
    while (resultsReceived < resultsExpected && cmd_number <= 9)
    {
        String command = "";
        command = "";
        command += sensorAddr;
        command += "D";
        command += cmd_number;
        command += "!";
        sensor->sendCommand(command);
        uint32_t start = millis();
        while (sensor->available() < 3 && (millis() - start) < 1500)
        {
        }
        sensor->read();          // ignore the repeated SDI12 address
        char c = sensor->peek(); 
        if (c == '+' || c == '-')
        {
            sensor->read();
        }

        while (sensor->available())
        {
            char c = sensor->peek();
            if (c == '-' || (c >= '0' && c <= '9') || c == '.')
            {
                float result = sensor->parseFloat(SKIP_NONE);
                Serial.print(", ");
                Serial.print(String(result, 10));
                mIndex[resultsReceived] = result;
                if (result != -9999)
                {
                    resultsReceived++;
                }
            }
            else
            {
                sensor->read();
            }
            delay(10); // 1 character ~ 7.5ms
        }
        cmd_number++;
    }
    Serial.println();
    measurementRequested = false;
}

include/ap200.hpp
#pragma once

#include <SDI12.h>

struct SensorInfos
{
    uint8_t addr;
    float sdiVersion;
    char vendorId[9];
    char model[7];
    float fwVersion;
    char serialNumber[14];
};

struct SensorMeasurements 
{
    float barometricPressure;
    float temperature;
    float ph;
    float oxydationReductionPotential;
    float electricalConductivity;
    float electricalConductivity20;
    float electricalConductivity25;
    float electricalResistivity;
    float salinity;
    float totalDissolvedSolid;
    float specificSeawaterGravity;
    float dissolvedOxygen;
    float dissolvedOxygenSaturation;
    float turbidity;
    float aux2;
    float nh3;
};

class Ap2000Sensor
{
private:
    SDI12* sensor;
    uint8_t sensorAddr;
    bool measurementRequested;
    bool IsSensorActive(uint8_t addr);                                  //Ping equivalent

public:
    Ap2000Sensor() = delete;
    Ap2000Sensor(const Ap2000Sensor&) = delete;
    void operator=(const Ap2000Sensor&) = delete;

    Ap2000Sensor(uint8_t sdiPin, uint8_t sensorAddress);
    ~Ap2000Sensor();

    SensorInfos GetSensorInfos();
    void DispSensorInfos(HardwareSerial& Serial);
    void DispSensorInfos(Serial_& Serial);
    void DispSensorInfos(HardwareSerial& Serial, SensorInfos& si);      //Spares execution time thus power
    void SensorBegin();
    void RequestMeasurements(uint8_t* mInfos);
    void RequestMeasurementsAndWaitUntilReady(uint8_t* mInfos);
    void RequestMeasurementsAndWaitUntilReady();
    void GetSensorMeasurements(SensorMeasurements* measurements,uint8_t resultsExpected = 16 , bool requestAgain = false);

    String units[16] = {"mBar", "°C", "ph", "mV", "uS/cm", "uS/cm", "uS/cm", "kO.cm", "PSU", "mg/L", "Sig_t", "mg/L", "%", "NTU", "none", "none"};
};


include/ap2000.hpp
#pragma once

/*!
 * @file ap2000.hpp
 *
 * @section intro_sec Introduction
 *
 * This is a library for Aquaread's AP-2000 sensor.
 *
 * @section author Author
 *
 * Written by The Last Resort.
 */


#include <SDI12.h>

struct SensorInfos
{
    uint8_t addr;
    float sdiVersion;
    char vendorId[9];
    char model[7];
    float fwVersion;
    char serialNumber[14];
};

struct SensorMeasurements 
{
    float barometricPressure;
    float temperature;
    float ph;
    float oxydationReductionPotential;
    float electricalConductivity;
    float electricalConductivity20;
    float electricalConductivity25;
    float electricalResistivity;
    float salinity;
    float totalDissolvedSolid;
    float specificSeawaterGravity;
    float dissolvedOxygen;
    float dissolvedOxygenSaturation;
    float turbidity;
    float aux2;
    float nh3;
};

class Ap2000Sensor
{
private:
    SDI12* sensor;
    uint8_t sensorAddr;
    bool measurementRequested;
    bool IsSensorActive(uint8_t addr);                                  //Ping equivalent

public:
    Ap2000Sensor() = delete;
    Ap2000Sensor(const Ap2000Sensor&) = delete;
    void operator=(const Ap2000Sensor&) = delete;

    Ap2000Sensor(uint8_t sdiPin, uint8_t sensorAddress);
    ~Ap2000Sensor();

    SensorInfos GetSensorInfos();
    void DispSensorInfos(HardwareSerial& Serial);
    void DispSensorInfos(Serial_& Serial);
    void DispSensorInfos(HardwareSerial& Serial, SensorInfos& si);      //Spares execution time thus power
    void SensorBegin();
    void RequestMeasurements(uint8_t* mInfos);
    void RequestMeasurementsAndWaitUntilReady(uint8_t* mInfos);
    void RequestMeasurementsAndWaitUntilReady();
    void GetSensorMeasurements(SensorMeasurements* measurements,uint8_t resultsExpected = 16 , bool requestAgain = false);

    String units[16] = {"mBar", "°C", "ph", "mV", "uS/cm", "uS/cm", "uS/cm", "kO.cm", "PSU", "mg/L", "Sig_t", "mg/L", "%", "NTU", "none", "none"};
};


src/ap2000.cpp
/*!
 * @file ap2000.cpp
 *
 * @section intro_sec Introduction
 *
 * This is a library for Aquaread's AP-2000 sensor.
 *
 * @section author Author
 *
 * Written by The Last Resort.
 */


#include "ap2000.hpp"


/*!
 *  @brief  Instantiates a new Ap2000Sensor class
 *  @param  sdiPin
 *          Pin number that sensor is connected
 *  @param  sensorAddress
 *          Address of the sensor
 */
Ap2000Sensor::Ap2000Sensor(uint8_t sdiPin, uint8_t sensorAddress)
{
    sensorAddr = sensorAddress;
    sensor = new SDI12(sdiPin);
}

/*!
 *  @brief  Destructor of the Ap2000Sensor class
 */
Ap2000Sensor::~Ap2000Sensor()
{
    delete sensor;
}

/*!
 *  @brief  Initiate the SDI-12 communication
 */
void Ap2000Sensor::SensorBegin()
{
    sensor->begin();
}

/*!
 *  @brief  Get the sensor general infos
 *
 *	@return Return all the general sensor infos
 */
SensorInfos Ap2000Sensor::GetSensorInfos()
{
    SensorInfos si;
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "I!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    si.addr = sensorReply.substring(0, 1).toInt();
    si.sdiVersion = sensorReply.substring(1, 3).toFloat() / 10;
    sensorReply.substring(3, 11).toCharArray(si.vendorId, 8);
    si.vendorId[8] = '\0';
    sensorReply.substring(11, 17).toCharArray(si.model, 6);
    si.model[6] = '\0';
    si.fwVersion = sensorReply.substring(17, 20).toFloat() / 100;
    sensorReply.substring(20).toCharArray(si.serialNumber, 13);
    si.serialNumber[13] = '\0';
    sensor->clearBuffer();

    return si;
}

/*!
 *  @brief  Get the sensor general infos and display them
 *  @param  serial
 *          Serial for framework that uses HardwareSerial
 */
void Ap2000Sensor::DispSensorInfos(HardwareSerial &serial)
{
    SensorInfos si = GetSensorInfos();

    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Display the sensor general infos provided
 *  @param  serial
 *          Serial for framework that uses HardwareSerial
 *  @param  si
 *          SensorInfos structure to display
 */
void Ap2000Sensor::DispSensorInfos(HardwareSerial &serial, SensorInfos &si)
{
    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Get the sensor general infos and display them
 *  @param  serial
 *          Serial for framework that uses Serial_
 */
void Ap2000Sensor::DispSensorInfos(Serial_& serial) {
    SensorInfos si = GetSensorInfos();

    serial.println("Sensor Address: " + static_cast<String>(si.addr));
    serial.println("Sensor SDI Version: " + static_cast<String>(si.sdiVersion));
    serial.println("Vendor Id: " + static_cast<String>(si.vendorId));
    serial.println("Sensor Model: " + static_cast<String>(si.model));
    serial.println("Sensor Firmware Version: " + static_cast<String>(si.fwVersion));
    serial.println("Sensor Serial Number: " + static_cast<String>(si.serialNumber));
}

/*!
 *  @brief  Ping as SDI-12 address
 *  @param  addr
 *          Address of the sensor to ping
 *  @return Return true if the sensor replied to the ping otherwise return false
 */
bool Ap2000Sensor::IsSensorActive(uint8_t addr)
{
    String cmd = (String)addr + '!';
    for (int i = 0; i < 3; i++)
    {
        sensor->sendCommand(cmd);
        delay(100);
        if (sensor->available())
        {
            sensor->clearBuffer();
            return true;
        }
    }
    sensor->clearBuffer();
    return false;
}

/*!
 *  @brief  Request measurments to the sensor
 *  @param  mInfos
 *          char[2] that get filled with the time to wait until the measurements are ready and how many there is
 */
void Ap2000Sensor::RequestMeasurements(uint8_t *mInfos)
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    mInfos[0] = sensorReply.substring(1, 4).toInt();
    mInfos[1] = sensorReply.substring(4).toInt();
    sensor->clearBuffer();
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 *  @param  mInfos
 *          char[2] that get filled with the time to wait until the measurements are ready and how many there is
 */
void Ap2000Sensor::RequestMeasurementsAndWaitUntilReady(uint8_t *mInfos)
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();

    mInfos[0] = sensorReply.substring(1, 4).toInt();
    mInfos[1] = sensorReply.substring(4).toInt();
    sensor->clearBuffer();
    delay(mInfos[0] * 1000);
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 */
void Ap2000Sensor::RequestMeasurementsAndWaitUntilReady()
{
    String sensorReply;

    sensor->sendCommand((String)sensorAddr + "C!");
    delay(100);
    sensorReply = sensor->readStringUntil('\n');
    sensorReply.trim();
    sensor->clearBuffer();

    delay(sensorReply.substring(1, 4).toInt() * 1000);
    measurementRequested = true;
}

/*!
 *  @brief  Request measurments to the sensor and wait until they are ready
 *  @param  measurements
 *          Structure that will be provisioned with the measurements received
 *  @param  resultsExpected
 *          The number of measurements expected
 *  @param  requestAgain
 *          Force a new measurement request before getting them
 */
void Ap2000Sensor::GetSensorMeasurements(SensorMeasurements* measurements, uint8_t resultsExpected, bool requestAgain)
{
    if (!measurementRequested || !requestAgain)
        this->RequestMeasurementsAndWaitUntilReady();

    uint8_t resultsReceived = 0;
    float* mIndex = (float*)measurements;
    uint8_t cmd_number = 0;
    while (resultsReceived < resultsExpected && cmd_number <= 9)
    {
        String command = "";
        command = "";
        command += sensorAddr;
        command += "D";
        command += cmd_number;
        command += "!";
        sensor->sendCommand(command);
        uint32_t start = millis();
        while (sensor->available() < 3 && (millis() - start) < 1500)
        {
        }
        sensor->read();          // ignore the repeated SDI12 address
        char c = sensor->peek(); 
        if (c == '+' || c == '-')
        {
            sensor->read();
        }

        while (sensor->available())
        {
            char c = sensor->peek();
            if (c == '-' || (c >= '0' && c <= '9') || c == '.')
            {
                float result = sensor->parseFloat(SKIP_NONE);
                Serial.print(", ");
                Serial.print(String(result, 10));
                mIndex[resultsReceived] = result;
                if (result != -9999)
                {
                    resultsReceived++;
                }
            }
            else
            {
                sensor->read();
            }
            delay(10); // 1 character ~ 7.5ms
        }
        cmd_number++;
    }
    Serial.println();
    measurementRequested = false;
}
